include/GafferScene/Private/IECoreScenePreview/Renderer.h

```diff
@@ -179,22 +193,24 @@ class Renderer : public IECore::RefCounted
 		/// "shutter", V2fData
 		/// The time interval for which the shutter is open - this is used in conjunction with the
 		/// times passed to motionBegin() to specify motion blur. Defaults to 0,0 if unspecified.
-		virtual ObjectInterfacePtr camera( const std::string &name, const IECore::Camera *camera ) = 0;
-
-		/// Adds a named light to the render. A shader for the light is expected to be
-		/// provided by a subsequently assigned attribute block in a renderer specific
-		/// fashion. Object may be non-NULL to specify arbitrary geometry for a geometric
-		/// area light, or NULL to indicate that the light shader specifies its own geometry
-		/// internally (or is non-geometric in nature).
-		virtual ObjectInterfacePtr light( const std::string &name, const IECore::Object *object = NULL ) = 0;
-
-		/// Adds a named object to the render.
+		virtual ObjectInterfacePtr camera( const std::string &name, const IECore::Camera *camera, const AttributesInterface *attributes ) = 0;
+
+		/// Adds a named light with the initially supplied set of attributes, which are expected
+		/// to provide at least a light shader. Object may be non-NULL to specify arbitrary geometry
+		/// for a geometric area light, or NULL to indicate that the light shader specifies its own
+		/// geometry internally (or is non-geometric in nature).
+		/// \todo Should object be typed as Primitive?
+		virtual ObjectInterfacePtr light( const std::string &name, const IECore::Object *object, const AttributesInterface *attributes ) = 0;
+
+		/// Adds a named object to the render with the initally supplied set of attributes.
+		/// The attributes may subsequently be edited in interactive mode using
+		/// ObjectInterface::attributes().
 		/// \todo Rejig class hierarchy so we can have something less generic than
 		/// Object here, but still pass CoordinateSystems. Or should
 		/// coordinate systems have their own dedicated calls?
-		virtual ObjectInterfacePtr object( const std::string &name, const IECore::Object *object ) = 0;
+		virtual ObjectInterfacePtr object( const std::string &name, const IECore::Object *object, const AttributesInterface *attributes ) = 0;
```

Comment:
I did consider that, but decided against it. I wanted to make it absolutely clear that it was necessary to provide the initial set of attributes (rather than subsequent edits) upfront and not after creation, and providing a default argument would make it possible to get that wrong.